### ISSUE TITLE
Per-thread overlay provider

### ISSUE THREAD

**Issue by Rjected** [paradigmxyz/reth#12688]
When computing the state root in parallel, we currently use the `ConsistentViewProvider`. When the persistence task is running, this would result in lots of failures due to the db moving. This is now expected, and if we have the following flow:
 * collect in-memory trie updates and anchor block
 * open database tx
 * make reverts with tx up to anchor block
 * construct overlay with reverts + in-memory updates

We can then use the overlay to create a provider that has a consistent view of the trie, even if the db moves while the provider is being used. This should then be used in the parallel root instead of the `ConsistentViewProvider`

**Comment by Rjected** [comment]
Note:

We've assumed:
> if the persistence is in progress and block that is currently being executed/validated is the descendant of the blocks that are being persisted - we can allow computing state root in parallel

We send single persistence actions synchronously, for example a remove action must complete before we send a save action. Save actions only ever extend the current chain.
If a reorg happens, and persistence is in progress, then we have the following cases:
* we're still saving some ancestor blocks, OR
* we're saving some non-ancestor blocks (we're calculating the state root on a sidechain) OR
* we're removing blocks.

If we're removing blocks or saving non-ancestor blocks:
* we do not run parallel state root.

If we're committing ancestor blocks, then:
Any trie updates being committed are a subset of the in-memory trie updates collected before fetching reverts. So it's possible that any diff in reverts (pre vs post commit) is already covered by the in-memory trie updates we collect.
